File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-java/parallel-consumer-proxy-client-java-api/src/main/java/bz/stub/parallelconsumer/client/Outcome.java
package bz.stub.parallelconsumer.client;
/*-
 * Copyright (C) 2026 Antony Stubbs and contributors
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

/**
 * The per-record verdict a {@link RecordProcessor} returns: success (optionally carrying records for Parallel
 * Consumer to produce) or failure (carrying a reason). Success lets the record's offset advance; failure sends
 * it back to Parallel Consumer's own retry scheduling, and the reason travels with the redelivery as
 * {@link InboundRecord#lastFailureReason()}.
 * <p>
 * A closed two-armed value rather than a class hierarchy, so a language with no subtyping or no exceptions
 * mirrors it directly (KTD1). There is deliberately no third arm: a processor that cannot decide has not
 * finished processing.
 *
 * @author Antony Stubbs
 */
public final class Outcome {

    private static final Outcome BARE_SUCCESS = new Outcome(true, Collections.emptyList(), null);

    private final boolean success;
    private final List<OutboundRecord> produce;
    private final String failureReason;

    private Outcome(boolean success, List<OutboundRecord> produce, String failureReason) {
        this.success = success;
        this.produce = produce;
        this.failureReason = failureReason;
    }

    /** Success with nothing to produce: the record's offset advances, and that is all. */
    public static Outcome success() {
        return BARE_SUCCESS;
    }

    /**
     * Success with records for Parallel Consumer to produce (R6) - the only sanctioned route for the
     * processor's Kafka output. An empty list is equivalent to {@link #success()}.
     */
    public static Outcome success(List<OutboundRecord> produce) {
        if (produce == null || produce.isEmpty()) {
            return BARE_SUCCESS;
        }
        return new Outcome(true, Collections.unmodifiableList(new ArrayList<>(produce)), null);
    }

    /**
     * Failure: the record returns to Parallel Consumer's retry scheduling, and the reason accompanies the
     * redelivery. The reason may be {@code null} when there is nothing useful to say.
     */
    public static Outcome failure(String reason) {
        return new Outcome(false, Collections.emptyList(), reason);
    }

    public boolean isSuccess() {
        return success;
    }

    /** The records to produce on success; always empty for a failure. */
    public List<OutboundRecord> produce() {
        return produce;
    }

    /** The failure reason; always empty for a success. */
    public Optional<String> failureReason() {
        return Optional.ofNullable(failureReason);
    }

    @Override
    public String toString() {
        return success ? "Outcome.success(produce=" + produce.size() + ")" : "Outcome.failure";
    }
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-typescript/src/outcome.ts
// Copyright (C) 2026 Antony Stubbs and contributors

import type { InboundRecord, OutboundRecord } from "./records";

/**
 * The per-record verdict: success (optionally carrying records to produce) or failure (carrying a
 * reason). A discriminated union rather than a class hierarchy - the TypeScript spelling of the
 * reference surface's closed two-armed value, narrowable by `outcome.kind` with no `instanceof`.
 *
 * There is deliberately no third arm. A processor that cannot decide has not finished processing,
 * and there is no per-record deadline that would let it stop: a record stays in flight until the
 * function returns, which is the product's differentiator rather than an oversight.
 */
export type Outcome =
  | { readonly kind: "success"; readonly produce: readonly OutboundRecord[] }
  | { readonly kind: "failure"; readonly reason: string | null };

/**
 * The user's function: one record in, one verdict out.
 *
 * THREE SPELLINGS, ONE MEANING, because TypeScript has three natural ones and forcing a single
 * spelling would be un-idiomatic in every direction:
 *
 * - return nothing - a bare success, the common case;
 * - throw, or reject - a failure, with the error's message as the reason. Exceptions ARE the
 *   language's error idiom, so a processor that lets one escape has said "failure" clearly;
 * - return an {@link Outcome} - needed when the success carries records to produce, or when a
 *   failure wants a reason without an exception.
 *
 * The translation happens in exactly one place ({@link applyProcessor}), so every spelling reaches
 * the wire through the same path.
 *
 * `async` is fully supported and expected: a processor returning a promise is awaited, and the
 * record stays in flight until it settles.
 */
export type RecordProcessor = (
  record: InboundRecord,
) => void | Outcome | Promise<void | Outcome>;

const BARE_SUCCESS: Outcome = { kind: "success", produce: [] };

/** Success, optionally with records for Parallel Consumer to produce. */
export function success(produce?: readonly OutboundRecord[]): Outcome {
  if (produce === undefined || produce.length === 0) {
    return BARE_SUCCESS;
  }
  return { kind: "success", produce: [...produce] };
}

/**
 * Failure: the record returns to Parallel Consumer's retry scheduling with its attempt count
 * consumed, and the reason travels with the redelivery as `lastFailureReason`.
 */
export function failure(reason?: string): Outcome {
  return { kind: "failure", reason: reason ?? null };
}

/**
 * Runs the processor and reduces every way it can end to one {@link Outcome}. The single place a
 * throw, a rejection, or a `void` return becomes a verdict.
 *
 * A processor that returns something that is not an outcome is a bug in the application, reported
 * as a failure rather than allowed to tear down the client: one bad record must not take the
 * session with it.
 */
export async function applyProcessor(
  processor: RecordProcessor,
  record: InboundRecord,
): Promise<Outcome> {
  let returned: void | Outcome;
  try {
    returned = await processor(record);
  } catch (thrown) {
    return failure(describe(thrown));
  }
  if (returned === undefined || returned === null) {
    return BARE_SUCCESS;
  }
  if (returned.kind === "success" || returned.kind === "failure") {
    return returned;
  }
  return failure("the processor returned a value that is not an Outcome");
}

/** An error's message, or its best textual form. Never a whole object dump. */
function describe(thrown: unknown): string {
  if (thrown instanceof Error) {
    return thrown.message.length > 0 ? thrown.message : thrown.name;
  }
  return typeof thrown === "string" ? thrown : "the processor threw a non-Error value";
}
